File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/secretsmanager/AwsSecretManager.java
package uk.nhs.nhsx.core.aws.secretsmanager;

import com.amazonaws.services.secretsmanager.AWSSecretsManager;
import com.amazonaws.services.secretsmanager.AWSSecretsManagerClientBuilder;
import com.amazonaws.services.secretsmanager.model.*;

import java.util.Optional;

public class AwsSecretManager implements SecretManager {

    private final AWSSecretsManager client = AWSSecretsManagerClientBuilder.standard()
            .build();

    @Override
    public Optional<SecretValue> getSecret(SecretName secretName) {
        try {
            GetSecretValueResult getSecretValueResult =
                client.getSecretValue(new GetSecretValueRequest().withSecretId(secretName.value));

            return Optional
                .ofNullable(getSecretValueResult.getSecretString())
                .map(SecretValue::of);

        } catch (AWSSecretsManagerException e) {
            return Optional.empty();
        }
    }
}

File: src/aws/lambdas/incremental_distribution/infra/src/main/kotlin/uk/nhs/nhsx/core/aws/secretsmanager/AwsSecretManager.kt
package uk.nhs.nhsx.core.aws.secretsmanager

import com.amazonaws.services.secretsmanager.AWSSecretsManager
import com.amazonaws.services.secretsmanager.model.GetSecretValueRequest
import com.amazonaws.services.secretsmanager.model.AWSSecretsManagerException
import java.lang.RuntimeException
import java.util.*

class AwsSecretManager(private val client: AWSSecretsManager) : SecretManager {

    override fun getSecret(secretName: SecretName): Optional<SecretValue> = try {
        Optional
            .ofNullable(client.getSecretValue(GetSecretValueRequest().withSecretId(secretName.value)))
            .map { it.secretString }
            .map(SecretValue::of)
    } catch (e: AWSSecretsManagerException) {
        Optional.empty()
    }

    override fun getSecretBinary(secretName: SecretName): ByteArray = try {
        val getSecretValueResult = client.getSecretValue(GetSecretValueRequest().withSecretId(secretName.value))
        val buffer = ByteArray(getSecretValueResult.secretBinary.remaining())
        getSecretValueResult.secretBinary[buffer]
        buffer
    } catch (e: AWSSecretsManagerException) {
        throw RuntimeException(e)
    }
}
